src/libraries/System.Private.CoreLib/src/System/IO/FileStreamImpl.cs

```diff
@@ -0,0 +1,648 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Runtime.Serialization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.IO
+{
+    internal sealed partial class FileStreamImpl : FileStreamStrategy
```

Comment:
In [this other PR](https://github.com/dotnet/runtime/pull/47085) doing a similar refactoring for `Random`, the `*Impl` classes are being declared inside the main `Random` class.

One advantage of doing that is that the `*Impl` classes are only visible to the main class. Do you know if there would be any other considerable benefit of following the same pattern in this PR? I know the classes are huge, so that could be a problem (unless it's possible to make them partial and define their body in a separate file).

Reply:
I would prefer to avoid having nested private types, as it would force us to have a very big partial class (`FileStream`) or an enormous single source file.

With partial types, it's not obvious where the implementation starts and ends and you always have to eyeball all `.cs` files to be 100% sure. Having them as separate types makes it also easier to ensure that the `Strategies` have no dependency to `FileStream` which has caused me some troubles (https://github.com/dotnet/runtime/pull/47128/commits/d30f3718aa1f3081de4074e5f419edf485a9c8d2)